crates/services/sync/src/service.rs

```diff
@@ -1,61 +1,198 @@
-use crate::Config;
-use fuel_core_types::services::p2p::BlockGossipData;
-use parking_lot::Mutex;
-use tokio::{
-    sync::{
-        mpsc,
-        oneshot,
+//! Service utilities for running fuel sync.
+use std::sync::Arc;
+
+use crate::{
+    import::{
+        Config,
+        Import,
+    },
+    ports::{
+        self,
+        BlockImporterPort,
+        ConsensusPort,
+        PeerToPeerPort,
     },
-    task::JoinHandle,
+    state::State,
+    sync::SyncHeights,
 };
 
-pub enum SyncStatus {
-    Stopped,
-    InitialSync,
-}
+use fuel_core_services::{
+    stream::{
+        BoxStream,
+        IntoBoxStream,
+    },
+    RunnableService,
+    RunnableTask,
+    Service,
+    ServiceRunner,
+    SharedMutex,
+    StateWatcher,
+};
+use fuel_core_types::blockchain::primitives::BlockHeight;
+use futures::StreamExt;
+use tokio::sync::Notify;
+
+#[cfg(test)]
+mod tests;
 
-pub enum SyncMpsc {
-    Status { ret: oneshot::Sender<SyncStatus> },
-    Start,
-    Stop,
+/// Creates an instance of runnable sync service.
+pub fn new_service<P, E, C>(
+    current_fuel_block_height: BlockHeight,
+    p2p: P,
+    executor: E,
+    consensus: C,
+    params: Config,
+) -> anyhow::Result<ServiceRunner<SyncTask<P, E, C>>>
+where
+    P: ports::PeerToPeerPort + Send + Sync + 'static,
+    E: ports::BlockImporterPort + Send + Sync + 'static,
+    C: ports::ConsensusPort + Send + Sync + 'static,
```

Comment:
Hmmm, do we need those constraints here? Maybe we can remove them like:
```suggestion
where
    P: ports::PeerToPeerPort,
```

Reply:
And for types down, we can remove all constraints, because we don't need them during creation of tasks=)

Reply:
@xgreenx we actually do because you can't add extra constraints to the `run` on `RunnableTask` so we are forced to put them on the type.

Reply:
You can add constraints on the `impl` section itself, we did that for all other services.